// grunt-connect-apimock/test/old/advancedMockformatDelay_test.js
'use strict';

var http = require('http');

/*
  ======== A Handy Little Nodeunit Reference ========
  https://github.com/caolan/nodeunit

  Test methods:
    test.expect(numAssertions)
    test.done()
  Test assertions:
    test.ok(value, [message])
    test.equal(actual, expected, [message])
    test.notEqual(actual, expected, [message])
    test.deepEqual(actual, expected, [message])
    test.notDeepEqual(actual, expected, [message])
    test.strictEqual(actual, expected, [message])
    test.notStrictEqual(actual, expected, [message])
    test.throws(block, [error], [message])
    test.doesNotThrow(block, [error], [message])
    test.ifError(value)
*/

var DELAY_TIME = 1000;

exports.connect_apimock = {
    setUp: function(done) {
        // setup here if necessary
        done();
    },

    no_delayed_response: function(test) {
        var starttime = new Date();
        test.expect(3);
        http.request({
            path: '/api/advanced/delay?foo=bar',
            method: 'GET',
            port: 8080
        }, function(response) {
            var data = '';
            response.on('data', function(chunk) {
                data += chunk;
            });
            response.on('end', function() {
                var endtime = new Date();
                var executionTime = endtime.getTime() - starttime.getTime();
                test.equal(response.statusCode, 200);
                var expected = '{"message":"default"}';
                test.equal(data, expected, 'It should return the default response');
                test.equal(executionTime >= DELAY_TIME, false, 'The response should not be delayed');
                test.done();
            });
        }).end();
    },
    delayed_response: function(test) {
        var starttime = new Date();
        test.expect(3);
        http.request({
            path: '/api/advanced/delay?foo=foo',
            method: 'GET',
            port: 8080
        }, function(response) {
            var data = '';
            response.on('data', function(chunk) {
                data += chunk;
            });
            response.on('end', function() {
                var endtime = new Date();
                var executionTime = endtime.getTime() - starttime.getTime();
                test.equal(response.statusCode, 200);
                var expected = '{"message":"delayed"}';
                test.equal(data, expected, 'It should return the delayed response');
                test.equal(executionTime >= DELAY_TIME, true, 'The response should be delayed');
                test.done();
            });
        }).end();
    },

};
